src/tools/ceph-dencoder/ceph_dencoder.cc

```diff
@@ -64,10 +69,25 @@ void usage(ostream &out)
 
 vector<DencoderPlugin> load_plugins()
 {
-  fs::path mod_dir{CEPH_DENC_MOD_DIR};
+  /*
+   * Make it to be compatible with vstart.sh.
+   *
+   * The vstart.sh will set the "plugin_dir" option to
+   * "$CEPH_BUILD_DIR/lib" and there has no "denc/" under it.
+   *
+   * If there has a "denc/" it should be from install path
+   * "${CEPH_INSTALL_FULL_PKGLIBDIR}/denc".
+   */
+  std::string plugin_dir = g_conf().get_val<std::string>("plugin_dir");
+  fs::path mod_dir{plugin_dir};
```

Comment:
i am not sure why `plugin_dir` is connected to the `denc` directory. could you help explain this to me?

Reply:
There are 4 denc libraries after building from source in `ceph/build/lib/`:
```
denc-mod-cephfs.so  denc-mod-common.so  denc-mod-osd.so  denc-mod-rbd.so  denc-mod-rgw.so
```
And then they will be installed to the `/usr/local/lib64/ceph/denc/` after you run `make install` or to `/usr/lib64/ceph/denc/` when you install ceph from rpm packages.

Currently the `ceph_dencoder` will always try to load the denc libraries from `${CEPH_INSTALL_FULL_PKGLIBDIR}/denc`. So if we won't install the binaries after building from source we will always get the following error:

```
[root@ceph build]# ./bin/rados --pool cephfs.a.data getxattr 10000002e14.00000000 parent | ./bin/ceph-dencoder type inode_backtrace_t import - decode dump_json
error getting xattr cephfs.a.data/10000002e14.00000000/parent: (2) No such file or directory
unable to load dencoders from "/usr/local/lib64/ceph/denc". it is not a directory.
class 'inode_backtrace_t' unknown
```

Unless we export an env manually by reading the source code:

```
export CEPH_LIB=/data/ceph/build/lib/
```

The `vstart.sh` will set the global `plugin_dir` to `/data/ceph/build/lib/` for us.

Reply:
From `options/`, the `plugin_dir` should be the path where the dynamically loaded plugins:

```
 720 - name: plugin_dir                                                                                                                                
 721   type: str
 722   level: advanced
 723   desc: Base directory for dynamically loaded plugins
 724   default: @CEPH_INSTALL_FULL_PKGLIBDIR@
 725   services:
 726   - mon
 727   - osd
 728   flags:
 729   - startup
```